### DISCUSSION TITLE
Experiences with using dioxus live views

### DISCUSSION THREAD

**Discussion by brianmay** [DioxusLabs/dioxus#3378] [General]
I thought I would share my experiences I have had with live views here. As somebody with experiences writing personal projects with Phoenix Live views, but now prefers a type safe language such as Rust over Elixir.

Code is at https://github.com/brianmay/phone_db_rust/tree/23ffda33ed86a2def8d62e4c90bf72134e5e69ac - but note it is in a state of flux. I have pointed to the exact version of my project which is likely to become old quickly.

I have looked only at the 0.6.0 release. I have no idea if these solutions are good or correct, but they "work for me". It is possible some of these have already changed.

References to bug reports of known related issues appreciated.

# Paths must start with a `/`

This was my first error. Turns out there is a problem with line here:

https://github.com/DioxusLabs/dioxus/blob/v0.6.0/packages/liveview/src/adapters/axum_adapter.rs#L48

```rust
let ws_path = format!("{}/ws", route.trim_start_matches('/'));
```

It will only work when the path is `/` otherwise it will generate an incorrect path.

```
| path | result | expected |
| /    | /ws    | /ws      |
| /xyz | xyz/ws | /xyz/ws  | 
```

I think multiple calls to `with_virtual_dom` are, for every component. So we end up with a websocket for every top level component. I think this is different from phoenix live views which is able to share a single websocket amongst multiple top level components.

This enabled phoenix to do rapid changes between top level components.

I note the master version is a bit different so this may already be resolved.

# trait defined for router without state

https://github.com/DioxusLabs/dioxus/blob/v0.6.0/packages/liveview/src/adapters/axum_adapter.rs#L36

```rust
impl LiveviewRouter for Router {
```

This makes it impossible to use on a `Router` that has called `with_state()`.

Tested fix:

```rust
impl MyRouter for Router<AppState> {
```

Actually this probably should be a generic.

# HTML Template code hard coded

https://github.com/DioxusLabs/dioxus/blob/v0.6.0/packages/liveview/src/adapters/axum_adapter.rs#L51-L62

```rust
        let index_page_with_glue = move |glue: &str| {
            Html(format!(
                r#"
        <!DOCTYPE html>
        <html>
            <head><title>{title}</title></head>
            <body><div id="main"></div></body>
            {glue}
        </html>
        "#,
            ))
        };
```

This may not always be suitable. e.g. it may be desirable to have my head elements. I have meta viewport.

I tried to generate this myself using rsx macros. This didn't work out though, I couldn't generate the `head` element from rsx, but also (perhaps not surprisingly) I couldn't convert the dom tree to a string for rendering. So I stack to a string based template as above.

https://github.com/brianmay/phone_db_rust/blob/23ffda33ed86a2def8d62e4c90bf72134e5e69ac/backend/src/lib.rs#L86-L103

# App State and Path parameters

My App needs access to the State and Path parameters. I could not see how to do this with the existing code. So I ended up replacing the `with_virtual_dom` with my own function.

https://github.com/brianmay/phone_db_rust/blob/23ffda33ed86a2def8d62e4c90bf72134e5e69ac/backend/src/lib.rs#L124-L181

This currently has hard-coded values for `AppState`, this could be changed to a generic parameter.

I have not attempted to deal with Query parameters, I think this would be next on my list.

With my changes, I can do do:

```rust
Router::new()
        .with_my_app("/contacts/:id", ContactDetailView)
```

where the `ContactDetailView` is defined like:

```rust
pub fn ContactDetailView(props: Props<i64>) -> Element {
```

And this will listen on `/contacts/:id` and `/contacts/:id/ws` respectively and the html path will pass the required information to the websocket path.

In hindsight, wondering if this is the best approach. Might be better to pass the raw information to the component and handle the parsing in the component. Or something like that.

I note the master version is a bit different, but I think it doesn't yet solve the issue.

# Websocket disruption

If websocket connection goes down, the website becomes non-responsive with no feedback to the user, and does not appear to try to connect again. This is not particularly good user experience.

# Redirection

I have an object detail page that has a delete operation. But this of course means the page will turn into an error because the object doesn't exist anymore.

Need to be able to redirect to a different page from the component, but I can't work out how.

# Infinite scrolling

Something I suspect is not currently possible without access to live scroll position and window size position information. Maybe this might help https://github.com/DioxusLabs/dioxus/issues/2128

I think not even wasm front end code provides the required access yet though, although presumably I could fall back to JavaScript here (I thought I saw an issue on this but can't find it right now).

# Conclusion

While live views work and seem to work well, I think there are currently some limitations. As described above.

Wondering if I should keep going with the live view approach, which is "good enough" for the limited audience my project (me) for now. Or if I should change to the wasm based front end, which I suspect might be more future proof.

I think the live views may require more changes to keep up with dioxus developments.

But before I even considered changing to wasm, I just wanted to document my experiences to this point. Which I have done here.

**Comment by ealmloff** [comment]
Thanks for taking the time to create this writeup! Some of these issues are fixed in the git version of dioxus, some could be fixed with better documentation, and some are more fundamental to the liveview approach. Unfortunately, Liveview is the only platform that handles events asynchronously which forces the API for communicating with the browser (through eval, onmounted, and event handlers) to be async. It also performs poorly with SEO and can be more resource intensive. Because of those limitations, we have been deprioritizing liveview as a platform and may drop support in the future. If we do drop support, some migration path for liveview-like apps will be available in fullstack like signals driven through websockets

> # Paths must start with a `/`

I think that should be fixed in https://github.com/DioxusLabs/dioxus/pull/3341

> # App State and Path parameters 

You can use the `Router` with liveview to parse the route and query under the `base_path` in your `Dioxus.toml` on the client. The git version of `with_virtual_dom` intercepts all paths under the route you pass in and parses the rest on the "client" websocket

> # HTML Template code hard coded

This is part of the reason dioxus in general is moving away from liveview. You can add elements to the head with the [`dioxus::document::*`](https://docs.rs/dioxus-document/latest/dioxus_document/index.html#functions) components after the websocket has loaded. If you care about SEO, fullstack prerenders everything on the server automatically unlike liveview

> # Redirection
> I have an object detail page that has a delete operation. But this of course means the page will turn into an error because the object doesn't exist anymore.
> 
> Need to be able to redirect to a different page from the component, but I can't work out how.

You can use the navigator with the router to change to a different page: https://dioxuslabs.com/learn/0.6/router/reference/navigation/programmatic

**Reply by brianmay** [reply]
> You can use the Router with liveview

Oh, OK. I assumed the router wouldn't work here. Guess I assumed wrongly :-)

I probably should try the Router, I think it will be required if I do switch to wasm anyway.

And it does sound like wasm front end is the way to go (full stack might be even better but I had issues with this that probably deserve separate topic).

I don't care about SEO, this is an internal only system.

Am kind of curious though that Phoenix developers seem to think that live view is a good approach and I would have thought Elixir would use more resources then a Rust approach. But maybe that is due to tradeoffs, and Elixir not having good WASM support so they consider it more important.

**Reply by brianmay** [reply]
For a Live View, I need to get access to that State<...> also. As that contains important information such as configuration and the db pool.

**Reply by brianmay** [reply]
I notice that #3341 has:

```rust
route = "/*route".to_string();
```

But I think this is insufficient, it won't match `/`.

From, the documentation at https://docs.rs/axum/latest/axum/struct.Router.html:

> Note that /*key doesn’t match empty segments. Thus:
>
> /*key doesn’t match / but does match /a, /a/, etc.
> /x/*key doesn’t match /x or /x/ but does match /x/a, /x/a/, etc.

So suspect it needs two routes, one for "/" and another one for "/*route".

**Comment by brianmay** [comment]
I find that there seems to be some confusion with life times I don't understand.

e.g. (cut down code)

```rust
pub fn PhoneCallListView(props: Props) -> Element {
    let mut edit_contact = use_signal(|| None);

    let mut new_phone_calls_signal: Signal<Vec<PhoneCallDetails>> = use_signal(std::vec::Vec::new);

    rsx! {
        for phone_call in new_phone_calls_signal.iter() {
            PhoneCallComponent {
                key: phone_call.id,
                show_actions: true,
                phone_call: phone_call.clone(),
                on_edit_contact: move |_contact_id| {
                    edit_contact.set(Some(phone_call.contact_id));
                }
            }
        }
    }
}
```

generates the error:

```
error[E0597]: `new_phone_calls_signal` does not live long enough
   --> backend/src/components/phone_calls.rs:169:27
    |
166 |       let mut new_phone_calls_signal: Signal<Vec<PhoneCallDetails>> = use_signal(std::vec::Vec::new);
    |           -------------------------- binding `new_phone_calls_signal` declared here
167 |
168 | /     rsx! {
169 | |         for phone_call in new_phone_calls_signal.iter() {
    | |                           ^^^^^^^^^^^^^^^^^^^^^^ borrowed value does not live long enough
170 | |             PhoneCallComponent {
171 | |                 key: phone_call.id,
...   |
178 | |         }
179 | |     }
    | |_____- argument requires that `new_phone_calls_signal` is borrowed for `'static`
180 |   }
    |   - `new_phone_calls_signal` dropped here while still borrowed
```

But If I remove the reference to `phone_call.contact_id` it works.

LATER: Oh, I see, I think there might be some sort of implied attempt of the rsx! macro (or maybe it is the nested component `PhoneCallComponent` that does it)  to claim ownership of `phone_call`, which won't work, as it is a reference.

I think I have to lookup the id value outside the rsx! macro like:

```rust
pub fn PhoneCallListView(props: Props) -> Element {
    let mut edit_contact = use_signal(|| None);

    let mut new_phone_calls_signal: Signal<Vec<PhoneCallDetails>> = use_signal(std::vec::Vec::new);

    rsx! {
        for phone_call in new_phone_calls_signal.iter() {
            {
                let id = phone_call.id;
                rsx! {
                    PhoneCallComponent {
                        key: phone_call.id,
                        show_actions: true,
                        phone_call: phone_call.clone(),
                        on_edit_contact: move |_contact_id| {
                            edit_contact.set(Some(id));
                        }
                    }
                }
            }
        }
    }
}
```

This isn't as elegant, but at least it works.

EVEN LATER: I have a `_contact_id` parameter, I should just use that. I am an idiot :-)

But will post this regardless, it is a good example of the life time issue I experienced.

**Comment by brianmay** [comment]
Does Dioxus full stack currently have any support for pushing data from the server to the client, e.g. via websockets?

I thought I saw something, but now I can't find it.

**Reply by brianmay** [reply]
I had a go at doing websockets.

https://github.com/brianmay/dioxus-fs

I am somewhat less then impressed at this solution.

Partly because it introduces extra complexity by having to launch the axum stuff manually.

Partly because I had to import axum types and gloo-net types that look similar but in actual fact are different. I guess I could ease this my spiting client stuff and server stuff into separate files.

And particular because while this works perfectly on Brave, for some reason on Firefox it can take a while to connect to the websocket. The API says it is connected, but it isn't, and sending to it hangs.

I haven't attempted to reconnect the websocket if the connection goes down.

**Reply by brianmay** [reply]
Oh, I suspect what might be happening with Firefox is it is rate limiting all websocket connections because connections to `ws://localhost:8080/_dioxus` are failing.

Lets see what happens if I add a dummy websocket handler for that.

Yes, that does seem to help.

**Reply by ealmloff** [reply]
Server functions don't support websockets yet, but they do support one way streaming like [this](https://github.com/DioxusLabs/dioxus/blob/main/examples/fullstack-streaming/src/main.rs#L26-L38) example. Websocket support is planned for a future release. Without server function support, you can still use a separate axum handler and client side connection like your example, but it is less ergonomic

**Reply by brianmay** [reply]
That is interesting...

But I can't actually test it. Unfortunately it requires a crate called `test_fn` which in turn conflicts with `dioxus::prelude::test_fn` when using `dioxus::prelude::*`.

And the `component` macro really wants to be able to refer to `test_fn`. which is now ambiguous.

So importing the elements from `dioxus::prelude` by hand does not fix the problem. Because now the macro will pick up the wrong reference.

**Reply by ealmloff** [reply]
> But I can't actually test it. Unfortunately it requires a crate called `test_fn` which in turn conflicts with `dioxus::prelude::test_fn` when using `dioxus::prelude::*`.

Could you share the code you are trying to compile? `dioxus::prelude::test_fn` doesn't exist in the git (or 0.6) version of dioxus as far as I know

**Reply by brianmay** [reply]
https://github.com/brianmay/dioxus-fs-demo

Reproduce the error with:

```
cargo add server_fn
cargo build --profile server-dev --verbose --features server --bin dioxus-fs-demo
```

No other changes required.

This gives me:

```console
E0659]: `server_fn` is ambiguous
   --> src/main.rs:373:1
    |
373 | #[server(EchoServer)]
    | ^^^^^^^^^^^^^^^^^^^^^ ambiguous name
    |
    = note: ambiguous because of a conflict between a name from a glob import and an outer scope during import or macro resolution
    = note: `server_fn` could refer to a crate passed with `--extern`
    = help: use `::server_fn` to refer to this crate unambiguously
note: `server_fn` could also refer to the crate imported here
   --> src/main.rs:22:5
    |
22  | use dioxus::prelude::*;
    |     ^^^^^^^^^^^^^^^^^^
    = help: consider adding an explicit import of `server_fn` to disambiguate
    = help: or use `crate::server_fn` to refer to this crate unambiguously
    = note: this error originates in the attribute macro `server` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0659]: `server_fn` is ambiguous
   --> src/main.rs:379:1
    |
379 | #[server(MagicNumber)]
    | ^^^^^^^^^^^^^^^^^^^^^^ ambiguous name
    |
    = note: ambiguous because of a conflict between a name from a glob import and an outer scope during import or macro resolution
    = note: `server_fn` could refer to a crate passed with `--extern`
    = help: use `::server_fn` to refer to this crate unambiguously
note: `server_fn` could also refer to the crate imported here
   --> src/main.rs:22:5
    |
22  | use dioxus::prelude::*;
    |     ^^^^^^^^^^^^^^^^^^
    = help: consider adding an explicit import of `server_fn` to disambiguate
    = help: or use `crate::server_fn` to refer to this crate unambiguously
    = note: this error originates in the attribute macro `server` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0659]: `server_fn` is ambiguous
   --> src/main.rs:385:1
    |
385 | #[server(GetPenguinEncounters)]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ambiguous name
    |
    = note: ambiguous because of a conflict between a name from a glob import and an outer scope during import or macro resolution
    = note: `server_fn` could refer to a crate passed with `--extern`
    = help: use `::server_fn` to refer to this crate unambiguously
note: `server_fn` could also refer to the crate imported here
   --> src/main.rs:22:5
    |
22  | use dioxus::prelude::*;
    |     ^^^^^^^^^^^^^^^^^^
    = help: consider adding an explicit import of `server_fn` to disambiguate
    = help: or use `crate::server_fn` to refer to this crate unambiguously
    = note: this error originates in the attribute macro `server` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0659]: `server_fn` is ambiguous
   --> src/main.rs:401:1
    |
401 | #[server(CreatePenguinEncounter)]
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ ambiguous name
    |
    = note: ambiguous because of a conflict between a name from a glob import and an outer scope during import or macro resolution
    = note: `server_fn` could refer to a crate passed with `--extern`
    = help: use `::server_fn` to refer to this crate unambiguously
note: `server_fn` could also refer to the crate imported here
   --> src/main.rs:22:5
    |
22  | use dioxus::prelude::*;
    |     ^^^^^^^^^^^^^^^^^^
    = help: consider adding an explicit import of `server_fn` to disambiguate
    = help: or use `crate::server_fn` to refer to this crate unambiguously
    = note: this error originates in the attribute macro `server` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0659`.
error: could not compile `dioxus-fs-demo` (bin "dioxus-fs-demo") due to 4 previous errors
```

**Reply by brianmay** [reply]
Also worth noting that the following line does not create any errors:

```rust
use dioxus::prelude::server_fn::error::NoCustomError;
```

**Reply by brianmay** [reply]
I created an issue for this. https://github.com/DioxusLabs/dioxus/issues/3438

**Reply by ealmloff** [reply]
You don't need to add `server_fn` as a dependency. dioxus includes it in the prelude

**Reply by brianmay** [reply]
OK.

Can confirm now that the example works.

Example is stupid, it ignores the result of the `unbounded_send` and keeps sending long after the client has gone. Easy to fix.

I guess that works by a simple HTTP feed that keeps sending data, instead of using a websocket. Is this correct?

**Reply by brianmay** [reply]
If I understand this correctly, is suitable for character stream only, not suitable as is for encoding/decoding json objects (because client not guaranteed to receive entire json object in one `next()` call).

Please correct me if I got this wrong :-)

**Reply by ealmloff** [reply]
https://github.com/DioxusLabs/dioxus/pull/3396 fixes the unbounded send error.

> If I understand this correctly, is suitable for character stream only, not suitable as is for encoding/decoding json objects (because client not guaranteed to receive entire json object in one next() call).

The [StreamingJson](https://docs.rs/server_fn/0.7.2/server_fn/codec/struct.StreamingJson.html) encoding uses serde_json to encode and decode each chunk of the stream into a concrete type. 

> Please correct me if I got this wrong :-)

HTTP streaming is chunked with a built in delimiter. It is not a single uninterrupted stream of binary data. More info [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding#chunked)

**Reply by brianmay** [reply]
I completely missed the `StreamingJson`. Will have to try that. Seems to require `JsonStream`, no surprise here.

What would also be interesting is to try to configure the client to automatically connect on start and reconnect if the connection goes down. I imagine it shouldn't be too hard, but I am use to unexpected problems :-)

websockets support still desirable, but this in theory should solve the problem for everything but my complicated MQTT based app (which I have no desire or time to port from yew anytime soon...)

**Comment by plcholder** [comment]
Whats the current state and experience of dioxus liveview now? Im really interested but see very little info

**Reply by brianmay** [reply]
LiveView is still around, and I believe it still works.

I was told that they intend to deprecate LiveView in the future in favor of full stack, but right now full stack has several issues that would exclude me from using in production code, e.g. #3550.

There is also #4079, but haven't reproduced that for a while, wonder if I just haven't triggered it recently, or maybe an updated Chrome browser somehow fixed that.

And #4066 was just weird.

**Reply by plcholder** [reply]
thanks for the info, it seems like its not ready Yet. I might move back to phoenix liveview or use something like leptos, I was excited about liveview capabilities with rust dx and binary size can throw it on any host, but its getting depreciated :(

**Reply by brianmay** [reply]
This is the post I was looking for: https://github.com/DioxusLabs/dioxus/discussions/3378#discussioncomment-11598713, See the paragraph "This is part of the reason dioxus in general is moving away from liveview..."

In general I am happy using full stack for my own personal websites, would be a bit nervous using it for large audience. I did just get a response to #3550 which I responded to. It is possible there is a better solution, I just haven't worked it out yet.
